### PR TITLE
Fix kubevirt alerts

### PR DESCRIPTION
**Pull request by Barakmor1**
**What this PR does / why we need it**:
the forllowing alerts:
'VirtOperatorRESTErrorsBurst'
'VirtOperatorRESTErrorsHigh'
'VirtApiRESTErrorsBurst'
'VirtApiRESTErrorsHigh'
'VirtControllerRESTErrorsHigh'
'VirtControllerRESTErrorsBurst'
'VirtHandlerRESTErrorsHigh'
'VirtHandlerRESTErrorsBurst'

should be triggered only when we have more than 5% or 80% of failed requests calls for the last hour or five minutes respectively.
Currently, when these alerts are triggered  they keeps firing even when we don't have failed requests anymore

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #
I ran script that create and remove 50 VMs waited a bit and repeated this process to get failed requests.  
comparison of this PR queries and the current queries: (the left graphs represent this PR queries):
**virt_handlers_failed_client_rest_requests_in_last_5m:**
image

as you can see with the current query the number  requests keep raising and never stops  although it should when we are speaking about requests in time intervals in that use-case.

**virt_handlers_client_rest_requests_in_last_5m:**
image

as you can see with the current query graph contain two pods to consider and that doesn't make sense because we should sum all virt-handler pod's requests.

**_UNIT TESTS:_**
image
the values represent the number of requests in 1 minute interval ,
as you can see in the old logic we got 0 new requests without error (code=200)
(speaking about values: "2 2 2 2 2 2 2 2 2 2")
and 0 requests with error (code=500/400) because the tests ran only for 5 minutes
and **we get the new requests with error (code=500/400) only after 9 minutes**
( speaking about values: "10 10 10 10 10 10 10 10 10 20" )
so that logic is obviously wrong i changed it to
values: '0+10x20' == values: "0 10 20 30 40 ... 190"
values: '0+100x20' == values :"0 100 200 .... 1900"
and change the test duration to 20 minutes
now the error condition(more than 80% of the requests has error code) is true for **_more than 5 minutes_** which should fire an alert

 other changes in the unit tests are very similar or aesthetic changes

**Special notes for your reviewer**:
I found the following links useful:
https://stackoverflow.com/review/late-answers/27275266
https://valyala.medium.com/how-to-optimize-promql-and-metricsql-queries-85a1b75bf986
https://prometheus.io/docs/prometheus/latest/configuration/unit_testing_rules/

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: t
Comment (dshchedr) [inline t:1]: Possibly this file added accidentally and should be removed?

File: pkg/virt-operator/resource/generate/components/prometheus.go
Comment (dshchedr) [inline pkg/virt-operator/resource/generate/components/prometheus.go:96]: to be honest, I don't understand why we need the namespace filter: `namespace='%s'`

File: pkg/virt-operator/resource/generate/components/util.go
Comment (dshchedr) [inline pkg/virt-operator/resource/generate/components/util.go:12]: hm, I think it can be simplified a bit. This expression shows the same numbers:

`clamp_min(sum(rest_client_requests_total{pod=~'virt-controller-.*'} - rest_client_requests_total{pod=~'virt-controller-.*'} offset 5m), 0)`

so in your case this might work:
`clamp_min(sum(%s - %s offset %dm), 0)`

File: pkg/virt-operator/resource/generate/components/prometheus.go
Comment (dshchedr) [inline pkg/virt-operator/resource/generate/components/prometheus.go:94]: curious, do we really need these additional metrics? If we are not going to use it anywhere else - may be it would be better to just update the logic for existing metrics

**Comment by dshchedr** [comment]
And one more question. May be it is out of scope of this PR, but related to the same alerts. For example:

https://github.com/kubevirt/kubevirt/blob/9ea669c75fe600bd7b44f57de7dfac4d5cc3dce1/pkg/virt-operator/resource/generate/components/prometheus.go#L279

These alerts will be triggered only when we have such status for 5 minutes.. Do we really need to wait for 5 minutes or we should get alert firing immediately?

File: t
Comment (Barakmor1) [reply t]: Yes , removed it Thanks!

File: pkg/virt-operator/resource/generate/components/prometheus.go
Comment (Barakmor1) [reply pkg/virt-operator/resource/generate/components/prometheus.go:96]: yes exactly(we used this filter before my modifications as well)

File: pkg/virt-operator/resource/generate/components/util.go
Comment (Barakmor1) [reply pkg/virt-operator/resource/generate/components/util.go]: There is a problem with this query because if we will get new kind of request in 5 minutes interval it won't be included,
For instance:

**time x** :  lets say we had the following requests vector:   {metricA 2, metricB 1}

and after 2 minutes we get new kind of request we can see it in metricC

so in **time x+5m** we will have the following metrics: {metricA 2 , metricB 1, **metricC 1**}

if we will use 'clamp_min(sum(%s - %s offset %dm), 0)' instead :

`sum( {metricA 2, metricB 1, metricC 1} - {metricA 2 , metricB 1}) = 0 `   **instead** of 1 which is the real result 

this is explained in the following link:
https://prometheus.io/docs/prometheus/latest/querying/operators/#arithmetic-binary-operators

"Between two instant vectors, a binary arithmetic operator is applied to each entry in the left-hand side vector and its matching element in the right-hand vector. The result is propagated into the result vector with the grouping labels becoming the output label set. The metric name is dropped. **Entries for which no matching entry in the right-hand vector can be found are not part of the result.**"

File: pkg/virt-operator/resource/generate/components/prometheus.go
Comment (Barakmor1) [reply pkg/virt-operator/resource/generate/components/prometheus.go:94]: Hey,
The motivations of defining these functions and records:
1)  `GeneratePromQlQueryForValueDifferenceInFiveMinutesInterval` is much more declarative and clearer than 
`clamp_min(sum(rest_client_requests_total{pod=~'virt-controller-.*'}) - sum(rest_client_requests_total{pod=~'virt-controller-.*')} offset 5m), 0)`.
2) if we will want to change the way we calculate the difference in the future with function we would have to modify only one line whereas without it we would have to modify 8 lines.
3) without functions we increase the chance to have typo in the future because of the IDE's code auto completion and the code won't compile if we would have a typo in the function name In contrast to string.

**Comment by Barakmor1** [comment]
> And one more question. May be it is out of scope of this PR, but related to the same alerts. For example:
> 
> https://github.com/kubevirt/kubevirt/blob/9ea669c75fe600bd7b44f57de7dfac4d5cc3dce1/pkg/virt-operator/resource/generate/components/prometheus.go#L279
> 
> These alerts will be triggered only when we have such status for 5 minutes.. Do we really need to wait for 5 minutes or we should get alert firing immediately?

yes totally there is no reason for us to wait 5 minutes in that case i agree, i removed the 5 minutes delay.

File: pkg/virt-operator/resource/generate/components/prometheus.go
Comment (iholder-redhat) [reply pkg/virt-operator/resource/generate/components/prometheus.go:94]: I agree that using a function is helpful here, but I also wonder if we need extra metrics.

What I would suggest is create another function instead of defining new metrics. For example, you can introduce a `generateAllRestRequestsQuery(pod, namespace string) string` function. This function, as its name suggests, can generate a query without adding another `Record` metric.

For example, calling `generateAllRestRequestsQuery("virt-operator", "my-namespace")` could return `"rest_client_requests_total{pod=~'virt-operator-.*', namespace=my-namespace}"`. Then, as @dshchedr said, you can only edit existing records by doing something like: `GeneratePromQlQueryForValueDifferenceInFiveMinutesInterval(generateAllRestRequestsQuery("virt-operator", "my-namespace"))`.

Comment (iholder-redhat) [reply pkg/virt-operator/resource/generate/components/prometheus.go:94]: of course, another function can be made for rest calls with errors.

File: pkg/virt-operator/resource/generate/components/util.go
Comment (iholder-redhat) [inline pkg/virt-operator/resource/generate/components/util.go:11]: nits: IMO the function names and parameters are very long, which makes it a bit hard to read. I would suggest:
`GeneratePromQlQueryForValueDifferenceInHourInterval()` -> `GenerateValueDifferenceForLastHour()`
`GeneratePromQlQueryForValueDifferenceInFiveMinutesInterval()` -> `GenerateValueDifferenceForLastMinute()`
`GeneratePromQlQueryForValueDifferenceInInterval()` -> `GenerateValueDifferenceForInterval()`
`metricsToSumUpAndCalculateDifferenceInInterval` -> `metric`

Comment (iholder-redhat) [reply pkg/virt-operator/resource/generate/components/util.go:11]: Also - please make all functions here private. I don't see a reason for them being public.

File: pkg/virt-operator/resource/generate/components/util.go
Comment (iholder-redhat) [inline pkg/virt-operator/resource/generate/components/util.go:12]: There are unneeded spaces here:
`"clamp_min(sum (%s ) - sum (%s offset %dm) , 0 )"` -> `"clamp_min(sum (%s) - sum(%s offset %dm), 0)"`

File: pkg/virt-operator/resource/generate/components/prometheus.go
Comment (iholder-redhat) [inline pkg/virt-operator/resource/generate/components/prometheus.go:231]: question: what happens if the for a minute we have very high quantity of REST calls which cause an alert to be triggered. What happens if after another 10 seconds all REST requests are back to normal and have very low failure rate. Does this mean that for 4 minutes we will see this alert firing over and over (since indeed in the last 5 minutes we had, on average, high amount of REST errors)?

If this is true then for 1 hour it'll be even worse - we would re-trigger the event over and over for an hour even if the problem is already solved.

I think that 5 minutes could be a bit extreme, but what about changing it to `3m` or something similar? Same goes for the hourly alerts of course.

**Review by iholder-redhat — COMMENTED** [review]
Great job!! Overall looks great to me, left some small comments.

File: pkg/virt-operator/resource/generate/components/util.go
Comment (Barakmor1) [reply pkg/virt-operator/resource/generate/components/util.go:11]: "of course, another function can be made for rest calls with errors." - I'm not sure about 'code' label's value and it's source but if you consider a case where different pods have different error codes we would have to create 2 function for each kind of pod.
also i tried to stay as much generic as possible and i think that this function might be useful for other kind of metrics such as number of migration in time interval and in that case we can't be sure that migration metric would have code label at all.

File: pkg/virt-operator/resource/generate/components/prometheus.go
Comment (Barakmor1) [reply pkg/virt-operator/resource/generate/components/prometheus.go:94]: yes your right it might be better to use variable or another function instead of records. 
Thanks!

File: pkg/virt-operator/resource/generate/components/prometheus.go
Comment (Barakmor1) [reply pkg/virt-operator/resource/generate/components/prometheus.go:231]: Hey Itamar,

if i understood correctly i think that firing is a state and that when we add `For:    "5m"` to the alert 
it means that Prometheus will fire an alert only if the condition (in the Expr's value) is true for 5m
so it would be a west of time to wait another 5 minutes when the error rate is high for at least one hour.
this is explained in here:

https://prometheus.io/docs/prometheus/latest/configuration/alerting_rules/

"The optional for clause causes Prometheus to wait for a certain duration between first encountering a new expression output vector element and counting an alert as firing for this element. In this case, **Prometheus will check that the alert continues to be active during each evaluation for 10 minutes before firing the alert. Elements that are active, but not firing yet, are in the pending state.**"

**Comment by Barakmor1** [comment]
/hold
I think that as Itamar and Denyis mentioned the usage of rule recording isn't good.
preaggregating every metric an application exposes is ANTI PARTTEN. 
While it is true that aggregation is a good idea to reduce cardinality for performance, it is counterproductive to overdo it.
aggregating metrics that are only used once is a waste of resources and would require unnecessary maintenance as metrics are added and removed over time. Instead, we should add aggregation as we need it in many places.

**Comment by Barakmor1** [comment]
/re/retest-required

File: pkg/virt-operator/resource/generate/components/util.go
Comment (Barakmor1) [reply pkg/virt-operator/resource/generate/components/util.go]: Done.
Thanks!

File: pkg/virt-operator/resource/generate/components/util.go
Comment (Barakmor1) [reply pkg/virt-operator/resource/generate/components/util.go]: Done.
Thanks!

File: pkg/virt-operator/resource/generate/components/prometheus.go
Comment (iholder-redhat) [reply pkg/virt-operator/resource/generate/components/prometheus.go:231]: Thanks for clarifying! :)

File: hack/prom-rule-ci/prom-rules-tests.yaml
Comment (iholder-redhat) [inline hack/prom-rule-ci/prom-rules-tests.yaml:335]: So to clarify, tell me if I got this right- the old test that's supposed to check for the percentage of bad requests in an hour actually simulated a run for 5 minutes? And it passed because our rules were wrong?

**Review by iholder-redhat — COMMENTED** [review]
Great work! Good job finding many flaws in our current Prometheus related code!

I'm not a Prometheus expert, but this looks good to me. Let's wait for an approval to verify everything here is ok.

@dshchedr PTAL.
/lgtm

File: hack/prom-rule-ci/prom-rules-tests.yaml
Comment (Barakmor1) [reply hack/prom-rule-ci/prom-rules-tests.yaml:335]: exactly :)

File: hack/prom-rule-ci/prom-rules-tests.yaml
Comment (enp0s3) [inline hack/prom-rule-ci/prom-rules-tests.yaml:13]: @Barakmor1 I know this maybe out of scope, but can you please add a comment that refers to this [link](https://github.com/prometheus/prometheus/blob/83a56e22ab2c51e79e125dad5d56107d3b0dd2fc/docs/configuration/unit_testing_rules.md#test-file-format) which explains the test file format?

File: pkg/virt-operator/resource/generate/components/prometheus.go
Comment (enp0s3) [inline pkg/virt-operator/resource/generate/components/prometheus.go:309]: @Barakmor1 IIUC here you are changing the functionality of `VirtOperatorRESTErrorsHigh`. What you wanted to achieve by removing the `For: "5m"` is an immediate alert, right? 
Isn't that required to be defined in another alert? because IIUC the aim of `VirtOperatorRESTErrorsHigh` is to alert about a constant failure of REST calls over time. For example having REST errors of 10-20 rate per sample for the whole 5 minutes.

File: hack/prom-rule-ci/prom-rules-tests.yaml
Comment (enp0s3) [inline hack/prom-rule-ci/prom-rules-tests.yaml:256]: @Barakmor1 I see that you have increased the amount from 2 to 20 but it's unclear to me why, maybe you can elaborate about this change in the commit message?

File: hack/prom-rule-ci/prom-rules-tests.yaml
Comment (enp0s3) [inline hack/prom-rule-ci/prom-rules-tests.yaml:258]: @Barakmor1 same here

**Review by enp0s3 — CHANGES_REQUESTED** [review]
@Barakmor1 Hi. Thank you! overall looks good, please see comments below

File: hack/prom-rule-ci/prom-rules-tests.yaml
Comment (Barakmor1) [reply hack/prom-rule-ci/prom-rules-tests.yaml:256]: Hey Igor ,
to be honest i changed the entire test because it simulated a case where alert shouldn't be fire hence it fail with correct logic.
 i will try to explain:
this is one of the current test use-cases :
```
         - interval: 1m
            input_series:
       ....
       - series: 'rest_client_requests_total{namespace="ci", pod="virt-operator-1", code="200"}'
        values: "2 2 2 2 2 2 2 2 2 2"
      - series: 'rest_client_requests_total{namespace="ci", pod="virt-operator-1", code="400"}'
        values: "10 10 10 10 10 10 10 10 10 20"
        ...
        - eval_time: 5m
        alertname: VirtOperatorRESTErrorsBurst
        exp_alerts:
          - exp_annotations:
              summary: "More than 80% of the rest calls failed in virt-operator for the last 5 minutes"
              runbook_url: "https://kubevirt.io/monitoring/runbooks/VirtOperatorRESTErrorsBurst"
            exp_labels:
              pod: "virt-operator-1"
              severity: "critical"
          ...
```
currently the test run only 5 minutes so the number of New `rest_client_requests_total{namespace="ci", pod="virt-operator-1", code="200"}` is 0 because it is always 2(no new requests)
and the number of new `rest_client_requests_total{namespace="ci", pod="virt-operator-1", code="400"}`
is also 0 because the test run for only 5 minutes.
so our alert condition should be 0/0 > 0.8 which should NOT fire an alert
actually when the test run for only five minutes it will never fire an alert because of the `For:   5m` in the alert configuration

File: hack/prom-rule-ci/prom-rules-tests.yaml
Comment (Barakmor1) [reply hack/prom-rule-ci/prom-rules-tests.yaml:13]: Done.

File: pkg/virt-operator/resource/generate/components/util.go
Comment (Barakmor1) [reply pkg/virt-operator/resource/generate/components/util.go:20]: Done.

File: pkg/virt-operator/resource/generate/components/prometheus.go
Comment (Barakmor1) [reply pkg/virt-operator/resource/generate/components/prometheus.go:309]: if i understood correctly i think that firing is a state and that when we add For: "5m" to the alert
it means that Prometheus will fire an alert only if the condition (in the Expr's value) is true for 5m
so it would be a west of time to wait another 5 minutes when the error rate is high for at least one hour.

**Review by enp0s3 — APPROVED** [review]
/approve
@Barakmor1 Thank you very much for the effort!

**Comment by iholder-redhat** [comment]
/lgtm
Great job!

**Pull request merged**
